Add IPFIX documentation

bcronje opened the pull request:
@enclave-alistair I hope this is in line with what you were expecting.

enclave-alistair commented on mkdocs.yml:
```diff
@@ -170,6 +170,7 @@ nav:
         - Introduction:  management/trust-requirements/index.md
         - User Authentication: management/trust-requirements/user-authentication.md
         - Public IP Address: management/trust-requirements/public-ip-address.md
+      - IPFIX: management/ipfix.md
```
Suggest we make this title less protocol-specific, perhaps "Network Metadata (IPFIX)"?

Open to alternatives, just feels like the menu item should be less protocol-focussed.

bcronje replied:
Good idea, I dont even think we need the (IPFIX) in there. Some alternatives:

1. Network Metadata (IPFIX)
2. Network Metadata
3. Network Analytics
4. Flow Analytics
5. Network Analysis
6. Flow Analysis
7. Network Reporting
8. Network Monitoring
9. ?

enclave-alistair replied:
I think there's a few places where we refer to it as "Flow Metadata". Should we call it that? Or "Network Flow Metadata"?

bcronje replied:
I think that's a good call. I suggest "Network Flow Metadata" as the navigation item, and headers to reinforce it's related to the network, and then in the actual doc it should be acceptable to use "network flow metadata" and just "flow metadata" interchangeably.

enclave-alistair replied:
Sounds good to me.

enclave-alistair commented on docs/management/ipfix.md:
```diff
@@ -0,0 +1,103 @@
+
+## Introduction
+
+Enclave implements an Internet Protocol Flow Information Export (IPFIX) exporter. IPFIX is an IETF protocol that provides a universal standard for the export and collection of IP flow information. The Enclave IPFIX exporter conforms to [RFC 7011](https://www.ietf.org/rfc/rfc7011.html) and [RFC 7012](https://www.ietf.org/rfc/rfc7012.html) and should be compatible with any IPFIX collector that supp ...
```
Would be handy to have a paragraph here indicating *why* we support IPFIX, i.e. words around network visibility and diagnostics.

enclave-alistair commented on docs/management/ipfix.md:
```diff
@@ -0,0 +1,103 @@
+
+## Step 3: Configure IPFIX
+
+IPFIX is configured using the JSON property `FlowMetadata`. Under `FlowMetadata` we provide the following configuration properties and options:
```
Do you think we should just generally call this feature "Flow Metadata" since that's how we refer to generally? Is that a good 'public' name?

enclave-alistair commented on docs/management/ipfix.md:
```diff
@@ -0,0 +1,103 @@
+
+![image](/images/management/windows-edit-profile.png)
+
+On Linux the Enclave configuration file is located at `/etc/enclave/profiles/Universe.profile` and can be edit using your preferred editor, e.g.
```
```suggestion
On Linux the Enclave configuration file is located at `/etc/enclave/profiles/Universe.profile` and can be edited using your preferred editor, e.g.
```

enclave-alistair commented on docs/management/ipfix.md:
```diff
@@ -0,0 +1,103 @@
+
+IPFIX is configured using the JSON property `FlowMetadata`. Under `FlowMetadata` we provide the following configuration properties and options:
+
+- `IpFixCollectors` - A list of zero of more IPFIX collector information in the string format "_Udp/Collector IP Address:Port_" e.g. `Udp/172.16.1.100:4739`. The example specifies that `UDP` is the transport protocol to use, `172.16.1.100` is the IP address of the IPFIX collector and the collector is listening on UDP port `4739`.
```
A list of zero more IPFIX collector 'endpoints' perhaps? Information feels like the wrong word/tense here.

enclave-alistair commented on docs/management/ipfix.md:
```diff
@@ -0,0 +1,103 @@
+You can specify multiple IPFIX collectors and Enclave will forward IPFIX flow information to all configured collectors. By default no exporters are configured.
+
+- `Direction` - Valid string values are `Both`, `SendingToPeer` and `ReceivedFromPeer`. The default value if not specified is `Both`.
+IPFIX supports single direction flow information per data record. This means a normal bidirectional flow e.g. a TCP session, will generate two IPFIX data records, one for ingress information and a second for egress information. This is the default behaviour and what will be expected when you configure `Both` as the direction. In certain circumstances this might not be ideal, imagine you are runni ...
```
Can we break this bullet point up into multiple paragraphs? Feels a bit like a wall of text.

enclave-alistair replied:
`In certain circumstances this might not be ideal, imagine`; use semicolon instead of comma -> `In certain circumstances this might not be ideal;`

enclave-alistair commented on docs/management/ipfix.md:
```diff
@@ -0,0 +1,103 @@
+
+- `Direction` - Valid string values are `Both`, `SendingToPeer` and `ReceivedFromPeer`. The default value if not specified is `Both`.
+IPFIX supports single direction flow information per data record. This means a normal bidirectional flow e.g. a TCP session, will generate two IPFIX data records, one for ingress information and a second for egress information. This is the default behaviour and what will be expected when you configure `Both` as the direction. In certain circumstances this might not be ideal, imagine you are runni ...
+> There is one exception to the direction setting. Frames dropped due to violation of [Access Control Rules](/management/policy/#access-control-rules) will always generate an IPFIX record and be exported, irrespective of the direction setting. For example, even if you have configure `ReceivedFromPeer`, if an outgoing frame to a remote Enclave peer is dropped due to an [Access Control Rules](/mana ...
```
Do we also report frames that flows that fail due a trust requirement that isn't met?

bcronje replied:
Yes we do, basically if the frame fails evaluation (that includes trust requirement and ACL checks) it will be reported. I will update to include trust requirements.

enclave-alistair commented on docs/management/ipfix.md:
```diff
@@ -0,0 +1,103 @@
+IPFIX supports single direction flow information per data record. This means a normal bidirectional flow e.g. a TCP session, will generate two IPFIX data records, one for ingress information and a second for egress information. This is the default behaviour and what will be expected when you configure `Both` as the direction. In certain circumstances this might not be ideal, imagine you are runni ...
+> There is one exception to the direction setting. Frames dropped due to violation of [Access Control Rules](/management/policy/#access-control-rules) will always generate an IPFIX record and be exported, irrespective of the direction setting. For example, even if you have configure `ReceivedFromPeer`, if an outgoing frame to a remote Enclave peer is dropped due to an [Access Control Rules](/mana ...
+
+- `Interval` - The interval in millisecond that Enclave will generate and export interim IPFIX flow records. If `Interval` is set to 0, then Enclave will not generate interim flow records, but will only generate and export flow records at the end of each flow. Interval should be either 0, or a multiple of 60000 (1 minute) as Enclave only processes interim records on a 1-minute frequency. The defa ...
```
```suggestion
- `Interval` - The interval in milliseconds at which Enclave will generate and export interim IPFIX flow records. If `Interval` is set to 0, then Enclave will not generate interim flow records, but will only generate and export flow records at the end of each flow. Interval should be either 0, or a multiple of 60000 (1 minute) as Enclave only processes interim records on a 1-minute frequency. The default value is 0.
```

enclave-alistair commented on docs/management/ipfix.md:
```diff
@@ -0,0 +1,103 @@
+> Note that if you have enabled IPFIX through this setting but have not configured a collector in `IpFixCollectors`, then the Enclave IPFIX exporter will effectively be disabled. You need to both set enabled to true and have a collector configured for the Enclave collector to function.
+
+
+To add the IPFIX configuration options, add the JSON settings at the bottom of the Enclave configuration file you edited in Step 2.
```
```suggestion
To add the IPFIX configuration options, add the JSON settings at the bottom of the Enclave configuration file you opened in Step 2.
```

enclave-alistair reviewed (changes requested):
Thanks Beyers, this looks about right. Requested a few additions/amendments, but generally this is at the right level.

bcronje commented:
@enclave-alistair I rewrote most of the introduction section to address the "why", also added an overview of the typical IPFIX architecture with more detail on collectors etc. Also applied all the other change requests. Note the bit about end-to-end encrypted overlays such as Enclave that makes it hard for traditional network device flow exporters like routers to effectively generate metadata for the underlying application traffic.

enclave-alistair commented on docs/management/network-flow-metadata.md:
```diff
@@ -0,0 +1,131 @@
+
+- `Direction` - Valid string values are `Both`, `SendingToPeer` and `ReceivedFromPeer`. The default value if not specified is `Both`.
+
+  IPFIX supports single direction flow information per data record. This means a normal bidirectional flow e.g. a TCP session, will generate two IPFIX data records, one for ingress information and a second for egress information. This is the default behaviour and what will be expected when you configure `Both` as the direction.
+
+  In certain circumstances this might not be ideal; imagine you are running Enclave on System A and System B, both have IPFIX enabled and is monitoring a TCP session between them. With the default `Both` direction, System A will export both ingress and egress data, while System B will also export both ingress and egress data. This means you end up with 4 flow records for a single TCP session. Wit ...
+
+  > There is one exception to the direction setting. Frames dropped due to violation of [Trust Requirements](/management/trust-requirements/) or [Access Control Rules](/management/policy/#access-control-rules) will always generate an IPFIX record and be exported, irrespective of the direction setting. For example, even if you have configure `ReceivedFromPeer`, if an outgoing frame to a remote Enc ...
```
You actually require one more level of indentation under the bullet point line to make the content render with the correct indentation.

```suggestion
    IPFIX supports single direction flow information per data record. This means a normal bidirectional flow e.g. a TCP session, will generate two IPFIX data records, one for ingress information and a second for egress information. This is the default behaviour and what will be expected when you configure `Both` as the direction.

    In certain circumstances this might not be ideal; imagine you are running Enclave on System A and System B, both have IPFIX enabled and is monitoring a TCP session between them. With the default `Both` direction, System A will export both ingress and egress data, while System B will also export both ingress and egress data. This means you end up with 4 flow records for a single TCP session. With the `SendingToPeer` and `ReceivedFromPeer` direction options you can limit flow information export to either egress information or ingress information respectively.

    > There is one exception to the direction setting. Frames dropped due to violation of [Trust Requirements](/management/trust-requirements/) or [Access Control Rules](/management/policy/#access-control-rules) will always generate an IPFIX record and be exported, irrespective of the direction setting. For example, even if you have configure `ReceivedFromPeer`, if an outgoing frame to a remote Enclave peer is dropped due to either a [Trust Requirements](/management/trust-requirements/) or an [Access Control Rules](/management/policy/#access-control-rules) violation, Enclave will still report on this egress record.
```

enclave-alistair reviewed (changes requested):
Great stuff, one markdown oddity to deal with, then we're good.

bcronje commented:
> Great stuff, one markdown oddity to deal with, then we're good.

Done, applied you change.

enclave-alistair reviewed (approved):
🚀
